\documentclass[12pt]{article}
\usepackage{tikz}
\usepackage[cm]{fullpage}
\usepackage{verbatim}
\usepackage{hyperref} 
\usepackage{graphicx}
\usepackage{natbib}
\usepackage{amsmath,amssymb}
\usepackage{dsfont}
\DeclareMathOperator*{\argmin}{arg\,min}
\DeclareMathOperator*{\sign}{sign}
\DeclareMathOperator*{\Lik}{Lik}
\DeclareMathOperator*{\Peaks}{Peaks}
\DeclareMathOperator*{\HotSpots}{HotSpots}
\newcommand{\Cost}{\text{Cost}}
\DeclareMathOperator*{\Diag}{Diag}
\DeclareMathOperator*{\TPR}{TPR}
\DeclareMathOperator*{\FPR}{FPR}
\DeclareMathOperator*{\argmax}{arg\,max}
\DeclareMathOperator*{\maximize}{maximize}
\DeclareMathOperator*{\minimize}{minimize}
\newcommand{\ZZ}{\mathbb Z}
\newcommand{\NN}{\mathbb N}
\newcommand{\RR}{\mathbb R}

\begin{document}

\title{Constrained maximum likelihood binomial segmentation for
  methylation count data}

\author{Toby Dylan Hocking}

\maketitle

\section{Introduction/problem}

Warren says that methylation sequencing data along the genome can be
represented as a pair $(t_j, m_j)$ where $t_j\in\ZZ_+$ is the total
number of reads at methylation site $j$ and $m_j\in\{0,1,\dots, t_j\}$
is the number of methylated reads. The goal in analyzing these data is
to classify the genome into hyper-methylated regions with high $m_j$
values, and hypo-methylated regions with low $m_j$ values.

\section{Binomial segmentation model}

We propose the model $m_j \sim \text{Binomial}(t_j, p_j)$. The
binomial distribution has the probability mass function
\begin{equation}
  \label{eq:binomial-prob}
  \text{Prob}(m, p) = \binom{t}{m} p^m (1-p)^{t-m}
\end{equation}
which means the loss function for site $j$ is
\begin{equation}
  \label{eq:binomial-loss}
  \ell_j(p) = (m_j - t_j) \log(1-p) - m_j\log p.
\end{equation}

\section{Standard dynamic programming algorithm}

Let $(j', j]$ denote the interval of indices including $j'+1$ up to
and including $j$. Let $t_{(j', j]} = \sum_{i=j'+1}^{j} t_i$ be the
cumulative sum of total reads on that interval; idem for $m_{(j',
  j]}$. The optimal probability of methylation on that segment is
$p_{(j', j]} = m_{(j', j]}/t_{(j', j]}$, and the optimal loss of that
segment is
\begin{equation}
  \label{eq:optimal-cost}
  c_{(j', j]} = 
  ( m_{(j', j]} - t_{(j', j]} )
  \log\left\{ 
    1 - p_{(j', j]}
  \right\}
  - m_{(j', j]}
  \log p_{(j', j]}.
\end{equation}

Let $\mathcal L_{s,j}$ be the optimal loss in $s$ segments up to data
point $j$. To compute $\mathcal L_{1, j}$ we first compute and store
the $d$-vectors of cumulative sums $m_{(0, j]}$ and $t_{(0, j]}$, for
all sites $j\in\{1, \dots, d\}$.

The DPA computes the optimal loss for all segments $s\in\{2, \dots,
s_{\text{max}} \}$ and for all sites $j\in\{1, \dots, d\}$ in
$O(s_{\text{max}} d^2)$ time. This is accomplished using the
recursion, for $s>1$,
\begin{equation}
  \label{eq:dpa-recursion}
  \mathcal L_{s, j} = 
  \min_{j' < j}
  \mathcal L_{s-1, j'}
  + 
  c_{(j', j]}.
\end{equation}
Note that the optimal cost $c_{(j', j]}$ may be computed in constant
$O(1)$ time given the $d$-vectors of cumulative sums $m_{(0, j]}$ and
$t_{(0, j]}$.

\section{Pruned dynamic programming algorithm}

To use the pDPA of \citet{pruned-dp} for the binomial distribution we
need to be able to find the solution of
\begin{equation}
  \label{eq:pDPA-intervals}
  a \log(1-p) + b \log p + c = d
\end{equation}
in terms of $p$, where $a,b$ are positive integers. Mathematica does
not provide a closed form solution. It can be re-written as
\begin{equation}
  \label{eq:pDPA-interval-factorized}
  e^{d-c} = (1-p)^a p^b,
\end{equation}
which can be further simplified (using the binomial theorem, since $a$
is a positive integer) to
\begin{equation}
  \label{eq:pDPA-interval-polynomial}
  e^{d-c} = \sum_{i=0}^a \binom{a}{i} (-1)^{a-i} p^{a-i+b},
\end{equation}
which makes it clear that the right side is a polynomial of degree
$a+b$. By the Abel-Ruffini theorem \citep{Abel-Ruffini-wikipedia}, if
the polynomial degree $a+b\geq 5$, then it is possible that there is
no closed-form solution in terms of radicals. However we can always
use an iterative root-finder such as Newton's method.

\includegraphics[width=\textwidth]{figure-binomial-loss}

The figure above shows the binomial loss function $\ell_j(p)$ for
$t_j\in\{10, 100\}$ (rows) and $m_j=\hat p t_j$ with $\hat p\in\{0.3,
0.5\}$ (columns). It is clear that the loss is convex, so there is
either 0, 1, or 2 intersection points with a constant function
(colored horizontal line). Finding the intersection points is a key
subroutine in the pDPA, and it seems that we will need to find the two
intersection points using a root-finder.

\bibliographystyle{abbrvnat}

\bibliography{refs}

\end{document}
